Reimplementação do Presto

PauloHMattos opened the pull request:
**[Proposito desse PR]:**
Organizar todo o projeto tornando-o orientado a objetos e mais modular, visando que o mesmo possa ser usado como base para a nova implementação do Trekking e de futuros seguidores de linha.

**[Principais mudanças]:**
- A IDE utilizada foi o Atom pois permite adicionar bibliotecas diretamente a pasta do projeto (compatibilidade com a Arduino IDE não garantida);
- Biblioteca _WheelEncoder_ implementada;
- Interface _MotorController_ definida: A partir dessa, implementou-se a classe _DifferentialDriveController_. Uma classe derivada da classe _DifferentialDriveController_, _PrestoMotorController_ foi criada interagir especificamente com o controlador de motor do Presto;
- Interface _InputSource_ definida, essa define qual a entrada da malha de controle do sistema: A partir dessa imlementou-se a classe _PrestoSensoring_ que produz a entrada do sistema a partir de seus sensores;
- Interface _SystemController_ definida: A partir dessa implementou-se a classe _PIDController_ e extendendo essa foi implementado _NonLinearPIDController_;
- Killswitch implementado;
- Biblioteca _QTRSensors_ atualizada para a versão 3.0.0;
- Arquivo de constantes (Constantes.h) removido uma vez que não há mais necessidade do mesmo;
- Foi implementado também dois filtros de low-pass para as leituras do array de sensores (_AverageFilter_ e _SimpleMovingAverageFilter_);
- Biblioteca de Logs implementada. Ela cria um buffer interno onde grava as mensagens, e uma vez por loop envia todo o buffer via Serial. Ao final do percurso o programa aguarda o apertar do botão pra emitir o relatório de execução (tempo que levou pra completar, constantes de PID, resultado do encoder, etc);
- Agora se usa uma função de ativação para transformar a velocidade de que saí do controle diferencial em PWM para o drive. Garante que os valores estejam sempre entre 0 e 1.

**[Limitações]:**
- Devido a como o Arduino trabalha com interrupções, só é possivel ter uma instância da classe _WheelEncoder_ por vez. Mas essa lida com 2 motores, e se necessário pode ser expandida para quantos precisar.

**[Situação de testes]:**
Aguardando montagem completa para os testes.

**[Risco técnico]:**
Alto. Todavia a nova abordagem tem um grau maior de organização e clareza no código.
Cada coisa está no seu devido lugar e o código em geral está mais simples de se manutenir.

**[Trabalho futuro]:**
- Remover a restrição das bibliotecas funcionarem apenas no Arduino (tornar o código portável para outros microcontroladores que suportem C++, como ARM). Isso envolve implementar uma biblioteca especifica de tempo, Log (substituí Serial), e possivelmente uma com métodos matemáticos básicos (abs, max, min, sin, cos, etc);
- Tornar possível extrair informações do Presto a cada tomada de tempo;
- Elaborar uma estratégia para tunar PID sem ser necessário reenviar o código a cada modificação;
- Melhorar a convenção dos nomes de variáveis e funções pra que se adequem melhor aos jargões da área de controle e robótica.

PauloHMattos committed:
Remake do código do presto

Organiza todo o projeto tornando-o orientado a objetos e mais modular,
visando que o mesmo possa ser usado como base para a nova implementação
do Trekking e de futuros seguidores de linha.

PauloHMattos committed:
Detalhes que eu esqueci no ultimo commit

PauloHMattos committed:
Merge branch 'master' into remake

lucianosantosdev reviewed (changes requested):
Existem algumas palavras reservadas a nivel de compilação que começam com _ ou __ e isso pode dar uma merdinha.

Em C++ sugiro trocar o "_" no início das variáveis privadas para "m_", é um padrão mais utilizado.

PauloHMattos commented:
Tô mais acostumado com C#, o padrão lá é esse. Vou mudar imediatamente.
Obrigado!

PauloHMattos committed:
Padronização do nome de variáveis e arquivos

As variaveis privadas, previamente declaradas com "_" como prefixo de
marcação de privacidade agora tem "m_" como prefixo para evitar
possiveis conflitos com variaveis internas do C++.

Os arquivos de header de classes de  também foram renomeados tendo seus
extenções modificadas ".h" (header de C) para ".hpp" (header de C++).
Isso pra tonar o arquivo mais claro sobre seu conteúdo, e auxiliar as
métricas do GitHub a informar a porcentagem de linguagem corretamente.
OBS: As bibliotecas _QTRSensors_ e _Button_ (que não foram implementadas
por nós) não foram modificadas para preservar seu estado original.

PauloHMattos committed:
Completa a implementação de Position

PauloHMattos committed:
Correção no encoder e controle de posição

PauloHMattos committed:
Controle dos motores do implementado

PauloHMattos committed:
Corrigido erro na biblioteca QTRSensors

A biblioteca dos sensores QTR da Pololu tem um erro na ordem das
operações para a leitura. Ele tenta deixar o estado de um pino digital
em HIGH (assumo que seja para carregar um capacitor), antes de mudar o
modo de operação para OUTPUT.

Não sei quão relevante isso é (dependendo da arquitetura do ATmega
talvez nem importe), e se já afetou de alguma forma o Presto. Todavia é
um erro de design e de fluxo do código, então foi corrigido.

PauloHMattos committed:
Faz a biblioteca de IR suportar o ATtiny45

PauloHMattos committed:
Update Presto_ATtiny_Encoder.ino

PauloHMattos committed:
Delete QTRSensors.cpp

PauloHMattos committed:
Delete QTRSensors.h

PauloHMattos committed:
Update WheelEncoder.hpp

Torna os contadores de ticks volatile pois são usados em interrupção.

Como instruído aqui: https://www.arduino.cc/en/reference/volatile

PauloHMattos committed:
Modifica a lógica do killswitch

Modifica o funcionando do killswitch pra que funcione direito com
interrupção

PauloHMattos committed:
Controle PID não linear

Implementa um controle por PID não linear semelhante ao que existia na
versão anterior.

Essa implementação foi feita como descrita num artigo da UFTPR presente
na pasta de referencias do período 2016.2.

A estrutura do PID também foi modificada pra facilitar implementar
variações desse controle. Os comentários foram levemente melhorados.

PauloHMattos committed:
Melhorias no controle de PID

- Habilita o controle de PID a rodar em tempos varíaveis;
- Corrige alguns erros de inicialização;
- Move algumas propriedades para a classe certa (SystemController ->
PIDController);
- Aprimora o Anti-ResetWindup (ARW) [1];
- Aprimora o calculo do fator integrativo [2];
- Melhora radicalmente a qualidade dos comentários :)

[1] A diferença é sutil, mas reduz o windup para zero em vez do erro
realmente pequeno que continuava a existir fazendo clamp do integrativo
pelos limites ao invés de fazer clamp do sinal de saída e corrigir o a
soma das integrais.

[2] Muda o algortimo de integração pra uma aproximação por sessões
trapezoidais. Resulta numa discretização do tempo mais aproximada para o
sistema.

PauloHMattos committed:
Move mais algumas coisas das classes de controle

PauloHMattos committed:
Low-pass filter p/ a leitura do array de sensores

Ele atua criando um buffer dos valores que foram lidos nas ultimas *n*
iterações, e calcula a média entre eles com valor da leitura atual.

Em seguida o valor mais antigo é substituido pelo mais recente.

Isso ameniza erros de leitura e evita que interfiram de forma
significativa no PID.

PauloHMattos committed:
Corrige o comportamento dos sensores de borda

A forma como estava implementada não considerava interseções da linha na
pista e faria com que o Presto parasse antes do fim da prova e/ou
reduzisse a velocidade sem motivo.

PauloHMattos committed:
LowPassFilter -> SimpleMovingAverageFilter

Apesar de ser um low-pass filter, o nome não descrevia muito bem do que
se tratava.
Agora o nome é bem especifico com o filtro implementado.

PauloHMattos committed:
Abstrai a lógica dos sensores

Talvez futuramente usemos outros tipos de filtro

PauloHMattos committed:
Corrige o filtro SMA

PauloHMattos committed:
Otimiza o calculo de PID não linear

A mudança evita que as constantes do calculo precisem ser computadas a
cada execução, economizando muito tempo de execução

PauloHMattos committed:
Update IRremoteInt.h

PauloHMattos committed:
Calibra o o valor pros sensores de borda automaticamente

PauloHMattos committed:
Pinos do IR para o kilswitch

Agora os pinos estão de acordo com a eletrônica do Presto

PauloHMattos committed:
Nova implementação do acionamento dos motores

PauloHMattos committed:
Inicio da biblioteca de matemática

PauloHMattos committed:
Configurações do PlatformIO

Corrige o tipo de placa nas configurações para não haver erros no upload

PauloHMattos committed:
Correção no PID não linear

PauloHMattos committed:
Correção de erro de compilação aleatório no encoder

PauloHMattos committed:
Biblioteca de Logs implementada

Prototipo pra biblioteca de logs. O objetivo é salvarmos tudo em um
buffer **dinamicamente alocado** (muito cuidado com isso), e depois que
o percurso for concluido plugamos em um computador pra receber tudo
atraves da porta serial

PauloHMattos committed:
Ajusta as classes de motor e sensores

Ajusta o tipo que recebe os pinos na função pra reduzir o uso de memória
e restringe o acesso ao set dos pinos.

Também diciona a LED 13 para debug visual exatamente como era no código
antigo

PauloHMattos committed:
Adiciona getter LineFollower::getIsRunning

PauloHMattos committed:
Arquivo para definir as diretivas de pré-compilamento

PauloHMattos committed:
Ajusta o código principal para as ultimas mudanças

PauloHMattos committed:
Simplifica um pouco o uso de PrestoSensoring

PauloHMattos committed:
Adiciona um novo modo de uso do filtro de média móvel simples

- Modo Continuo (ou de Histórico) faz a média das ultimas
'm_SamplesCapacity' leituras para amenizar ruido. Extremamente rápido
mas pode causar lag devido a variações nas ultimas leituras. Pode também
propagar rúido das ultimas avaliações

- Modo Estático, realiza a leitura 'm_SamplesCapacity' vezes e então faz
a média disso. É mais lento mas evita propagar erros e teoricamente não
apresenta lag de histórico.

Simplifica também o uso do filtro.

PauloHMattos committed:
Corrige a lógica de SimpleMovingFilter

Adiciona comentários também

PauloHMattos committed:
Corrige o Logger e separa os filtros

Pra que os nomes façam sentido eu separei os dois modos de operação que
eu tinha feito em 2 filtros separados.

Também foi corrigido a biblioteca de Logs que não estava nem perto de
certa

PauloHMattos committed:
Adiciona comentários e logs

PauloHMattos committed:
Sistema de comandos serial

Habilita modificarmos alguns parâmetros do código como constantes do PID
e velocidade linear máxima enviando comandos via serial.

Pra isso basta lugar o shield de programação e usar o Monitor Serial da
IDE da Arduino IDE.

A ideia é não perdermos tempo recompilando e enviarando o código
enquanto mexemos nesses valores durantes os testes. No log de fim de
prova o Presto enviar quais os valores foram utilizados.

Também diminui o tamanho do buffer de logs. É um sistema embarcado com
pouca RAM e eu já tô abusando um pouco de alocações dinâmicas no filtro
e no Log. Dê 1024 (50% da RAM do atmega328) foi pra 256 bytes. Isso deve
ser mais que suficiente.

PauloHMattos committed:
Funções de transferência de velocidade -> PWM

Define funções de ativação (ou transferência) que converte a velocidade
em cada roda para um valor normalizado entre 0 e 1 que depois é
multiplicado por uma PWM máxima.

Padroniza também os comentários no código (// -> /**/).

PauloHMattos committed:
Comentários sobre o trekking de posição

PauloHMattos committed:
Corrige erro de lógica no PID

PauloHMattos committed:
Corrige o uso de diretivas de pré-compilação

PauloHMattos committed:
Ajeita nome de função e padrão de comentários

PauloHMattos committed:
(Acho que) corrige que um erro no filtro de média móvel

PauloHMattos committed:
Refaz a lógica de BufferLogger

PauloHMattos committed:
Atualiza a biblioteca QTRSensors

- Atualiza pra ultima versão no Github da Pololu;
- Corrigido bug quando uma leitura bruta é menor do que a leitura de
calibração mínima que. Faz com que o valor salta para 1023;
- Corrigido erro de ordem ao carregar o capacitor do sensor.

PauloHMattos committed:
Configura novas regras de parada

Baseado no número de marcações a direita, e no tempo máximo de prova

PauloHMattos committed:
Corrige a regra de parada

O valor deve ser em milissegundos e não em segundos

PauloHMattos committed:
Implementado um filtro de transição

Semelhante ao que existia no código antigo, mas ele pode aceitar várias
funções.

No momento só tem a linear, como era no código antigo.

PauloHMattos committed:
Corrige a inicialização das arrays

PauloHMattos committed:
Habilita o Presto a ser executado várias vezes sem precisar desligar

PauloHMattos committed:
Impede que os motores sejam revertidos

Força o movimento sempre para frente por não admitir velocidades menores
que zero.
Caso a velocidade de algum dos motores seja calculada como menor que
zero a mesma é igualada a zero, fazendo com que a roda ligada a esse
motor se torne o eixo de rotação do Presto.

agathapires committed:
Ajusta o killswitch como o testado em 'master'

Define também o pino do encoder

PauloHMattos committed:
Corrige a implementação do filtro de média móvel

PauloHMattos committed:
Simplifica o controle do motor do Presto

E corrige um erro na função de transformação (velocidade -> PWM)

PauloHMattos committed:
Corrige a pinagem do attiny do encoder

PauloHMattos committed:
Proteção de inclusão de Definitions.h

PauloHMattos committed:
Modifica a classe sensoriamento do Presto

Com essas mudanças a biblioteca QTRSensors se torna desnecessário
portanto foi removida.

Com isso o único código presente do Presto que não foi desenvolvido pela
equipe é a biblioteca Button, que é totalmente desnecessária, e
futuramente será removida.

A nova classe de sensoriamento ainda requer algumas otimizações, mas
apenas de olho ela já deveria ter uma performance melhor pois utiliza
menos loops e operações aritméticas.

Além disso modifiquei a extensão do arquivo principal de .ino para .cpp.
A ideia é evitar que o Atom precise converter o arquivo a cada build.

PauloHMattos committed:
Implementa de novo o controle dos motores

Agora segue as especificações que planejamos para o novo shield em que o
driver será ligado.

O esquema que foi montado permite controlar a PWM e o sentido de rotação
usando apenas 2 pinos por motor.

PauloHMattos committed:
Controle de motor e encoder

PauloHMattos committed:
Remoção de variável não usada

PauloHMattos committed:
Código dos testes realizados no dia 24/06/2017

OBS: IR parou de funcionar de uma hora para outra sem nenhum motivo
aparente.

OBS²: Sem IR não consegui testar tanto quanto queria mas comecei a
ajustar as constantes do PID. Ele tá seguindo bem a linha em retas, mas
exagera na correção nas curvas, e por isso saí delas. Vou pro Fundão
amanhã continuar nisso e ver se descubro o que aconteceu com o IR

PauloHMattos committed:
Testes realizados no dia 26/06

Começou a seguir, mas ainda tá bem instável.

rkatopodis committed:
New PID tunings

Lucas-Malek committed:
Velocidade variante com o erro e volta completa.

Limpei um pouco o código também

PauloHMattos committed:
Constantes do PID.

Testes dia 01/07/2017

PauloHMattos committed:
Testes dia 02/07

PauloHMattos committed:
Update Presto.cpp

PauloHMattos committed:
Update LineFollower.cpp

rkatopodis committed:
Initial killswitch implementation using the bluetooth module

rkatopodis committed:
Bug fix
